dotnet/WPF/VisualLayerIntegration/VisualLayerIntegration/BarGraphUtility/BarGraph.cs

```diff
@@ -0,0 +1,479 @@
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//  ---------------------------------------------------------------------------------
+
+using SharpDX;
+using SharpDX.DirectWrite;
+using SharpDX.Direct2D1;
+using TextAntialiasMode = SharpDX.Direct2D1.TextAntialiasMode;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using SysWin = System.Windows;
+using Windows.UI;
+using Windows.UI.Composition;
+
+namespace BarGraphUtility
+{
+    class BarGraph
+    {
+        private float[] graphData;
+
+        private Compositor compositor;
+        private IntPtr hwnd;
+
+        private float graphWidth, graphHeight;
+        private float shapeGraphContainerHeight, shapeGraphContainerWidth, shapeGraphOffsetY, shapeGraphOffsetX;
+        private float barWidth, barSpacing;
+        private float maxBarValue;
+
+        private GraphBarStyle graphBarStyle;
+        private List<Windows.UI.Color> graphBarColors;
+
+        private Hashtable barValueMap;
+
+        private WindowRenderTarget textRenderTarget;
+        private SolidColorBrush textSceneColorBrush;
+        private TextFormat textFormatTitle;
+        private TextFormat textFormatHorizontal;
+        private TextFormat textFormatVertical;
+
+        private ShapeVisual shapeContainer;
+        private CompositionLineGeometry xAxisLine;
+        private CompositionLineGeometry yAxisLine;
+        private ContainerVisual mainContainer;
+
+        private static SharpDX.Mathematics.Interop.RawColor4 black = new SharpDX.Mathematics.Interop.RawColor4(0, 0, 0, 255);
+        private static SharpDX.Mathematics.Interop.RawColor4 white = new SharpDX.Mathematics.Interop.RawColor4(255, 255, 255, 255);
+
+        private static float textSize = 20.0f;
+
+        private AmbientLight ambientLight;
+        private SpotLight barOutlineLight;
+        private PointLight barLight;
+
+        #region public setters
+        public string Title { get; set; }
+        public string XAxisLabel { get; set; }
+        public string YAxisLabel { get; set; }
+        public ContainerVisual BarRoot { get; }
+        public ContainerVisual GraphRoot { get; }
+        #endregion
+
+        public enum GraphBarStyle
+        {
+            Single = 0,
+            Random = 1,
+            PerBarLinearGradient = 3,
+            AmbientAnimatingPerBarLinearGradient = 4
+        }
+
+        // Constructor for bar graph.
+        // To insert graph, call the constructor then use barGraph.Root to get the container to parent.
+        public BarGraph(Compositor compositor, IntPtr hwnd, string title, string xAxisLabel, 
+            string yAxisLabel, float width, float height, double dpiX, double dpiY, float[] data, WindowRenderTarget renderTarget, 
+            bool AnimationsOn = true, GraphBarStyle graphBarStyle = GraphBarStyle.Single,
+            List<Windows.UI.Color> barColors = null)
```

Comment:
Use an array unless you expect the callee to modify the colors.

Reply:
updated